test/library/library.go

```diff
@@ -32,3 +47,117 @@ func GenerateNameForTest(t *testing.T, prefix string) string {
 	// ensure no `-` at beginning or end
 	return strings.Trim(string(name), "-")
 }
+
+// WaitForAPIServerToBeOnTheSameRevision waits until all API Servers are running at the same revision.
+// The API Servers must stay on the same revision for at least waitForAPIRevisionSuccessThreshold * waitForAPIRevisionSuccessInterval
+//
+// Observations:
+//  rolling out a new version is not instant you need to account for a propagation time (~1/2 minutes)
+//  rolling out a new version can take ~10 minutes
+func WaitForAPIServerToBeOnTheSameRevision(t *testing.T, podClient corev1client.PodInterface) {
+	if err := wait.Poll(waitForAPIRevisionPollInterval, waitForAPIRevisionTimeout, mustSucceedMultipleTimes(waitForAPIRevisionSuccessThreshold, waitForAPIRevisionSuccessInterval, func() (bool, error) {
+		converged, err := areAPIServersOnTheSameRevision(podClient)
+		return converged, err
+	})); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// areAPIServersOnTheSameRevision tries to find the current revision that the API servers are running at.
+// The number of instances is calculated based on the number of running pods in a namespace.
+// This should be okay because this function is meant to be used by WaitForAPIServerToBeOnTheSameRevision which will wait at least waitForAPIRevisionSuccessThreshold * waitForAPIRevisionSuccessInterval
+// The number of pods should stabilize in that period of time.
+func areAPIServersOnTheSameRevision(podClient corev1client.PodInterface) (bool, error) {
+	revisionLabel := "revision"
+
+	// do a live list so we never get confused about what revision we are on
+	apiServerPods, err := podClient.List(context.TODO(), metav1.ListOptions{LabelSelector: "apiserver=true"})
+	if err != nil {
+		return false, err
+	}
+
+	good, bad, progressing, err := categorizePods(apiServerPods.Items)
+	if err != nil {
+		return false, err
+	}
+	if progressing {
+		return false, nil
+	}
+
+	goodRevisions := revisions(revisionLabel, good)
+	failingRevisions := revisions(revisionLabel, bad)
+
+	if len(goodRevisions) != 1 {
+		return false, nil // api servers have not converged onto a single revision
+	}
+	revision, _ := goodRevisions.PopAny()
+	if len(revision) == 0 {
+		revision = "0"
+	}
+
+	if failingRevisions.Has(revision) {
+		return false, fmt.Errorf("api server revision %s has both running and failed pods", revision)
+	}
+
+	return true, nil
+}
+
+func revisions(revisionLabel string, pods []*corev1.Pod) sets.String {
+	ret := sets.NewString()
+	for _, p := range pods {
+		ret.Insert(p.Labels[revisionLabel])
+	}
+	return ret
+}
+
+func categorizePods(pods []corev1.Pod) (good []*corev1.Pod, bad []*corev1.Pod, progressing bool, err error) {
```

Comment:
I didn't clue in to the majority of these additions being copied from library-go until after I'd reviewed this method in detail. Given that this function has to be maintained separately, though, I have some suggestions for simplification/cleanup:

- Rather than use a mix of named and literal return values, maybe pick one approach or the other and stick with it?
- Maybe collect sets of good and bad revisions directly rather than returning pods? The original implementation returned pods because it needed both revisions and node names but that's not a requirement here.
- Maybe treat both sets of revisions being empty as an implicit indication of 'progressing' and skip the explicit return value?

Reply:
I like all your suggestion except for the last one. I think we might have a mix of pods in `PodRunning` and `PodPending` state which indicate that we are progressing.

Reply:
> if err is nil and both good and bad revision sets are nil -> keep trying

Mhm, I don't see when `getRevisions` could return the above . . .

Reply:
This is exactly the case where `progressing=true` is returned. i.e this function is returning a boolean that could be inferred.